mllam_data_prep/config.py

```diff
@@ -151,13 +169,20 @@ class InputDataset:
         (e.g. two datasets that coincide in space and time will only differ in the feature dimension,
         so the two will be combined by concatenating along the feature dimension).
         If a single shared coordinate cannot be found then an exception will be raised.
+    derived_variables: Dict[str, DerivedVariable]
+        Dictionary of variables to derive from the dataset, where the keys are the variable names and
+        the values are dictionaries defining the necessary function and kwargs. E.g.
+        `{"toa_radiation": {"kwargs": {"time": "time", "lat": "lat", "lon": "lon"}, "function": "calculate_toa_radiation"}}`
+        would derive the "toa_radiation" variable using the `calculate_toa_radiation` function, which
+        takes `time`, `lat` and `lon` as arguments.
     """
 
     path: str
     dims: List[str]
-    variables: Union[List[str], Dict[str, Dict[str, ValueSelection]]]
     dim_mapping: Dict[str, DimMapping]
     target_output_variable: str
+    variables: Union[List[str], Dict[str, Dict[str, ValueSelection]]] = None
+    derived_variables: Dict[str, DerivedVariable] = None
```

Comment:
Since we now can have either `variables` or `derived_variables`, `variables` can no longer be required and has to be optional. I couldn't figure out a way to specify that we need to require either `variables`  or `derived_variables` and that both cannot be missing. Any ideas?

Reply:
That is a good point. I think it is fine to make them both optional here in the spec and then we can enforce this when we load the config :) So probably we should create a `load_config()` function that for now wraps `Config.from_yaml_file()` (https://github.com/ealerskans/mllam-data-prep/blob/feature/derive_forcings/mllam_data_prep/create_dataset.py#L286). This kind of check can't be done at the type level (which is what dataclasses prescribe) but one would have to use something like [pydantic](https://docs.pydantic.dev/latest/) for that. I haven't added use of pydantic yet because until now I didn't think we needed it and I wanted to add yet another dependency. But we could add that in a separate PR. For now I would just wrap the `Config.from_yaml_file()` call, you could maybe include an inline comment to say that this check could eventually be done with pydantic.

Reply:
It sounds like a good idea, but I don't really know how it should be done :)

Reply:
Thanks, I will try to implement this!

Reply:
`variables` and `derived_variables` are no longer mutually exclusive - we can have both in the same dataset now :)
I have taken inspiration from your solution @mafdmi and added a check which requires us to have defined at least one of `variables` and `derived_variables` and also a check so that they both don't contain the same variable names.